src/util-port-interval-tree.c

```diff
@@ -0,0 +1,325 @@
+    int stack_size = STACK_SIZE;
+
+    while (current || stack_depth) {
+        while (current != NULL) {
+            if (current->max < port) {
+                current = NULL;
+                break;
+            }
+            const bool is_overlap = IsOverlap(port, port2, current);
+
+            if (is_overlap && (new_port == NULL)) {
+                /* Allocate memory for port obj only if it's first overlap */
+                new_port = DetectPortInit();
+                if (new_port == NULL) {
+                    goto error;
+                }
+
+                SCLogDebug("Found overlaps for [%u:%u], creating new port", port, port2);
+                new_port->port = port;
+                new_port->port2 = port2;
+                SigGroupHeadCopySigs(de_ctx, current->sh, &new_port->sh);
+
+                /* Since it is guaranteed that the ports received by this stage
+                 * will be sorted, insert any new ports to the end of the list
+                 * and avoid walking the entire list */
+                if (*list == NULL) {
+                    *list = new_port;
+                    (*list)->last = new_port;
+                } else if (((*list)->last->port != new_port->port) &&
+                           ((*list)->last->port2 != new_port->port)) {
+                    DEBUG_VALIDATE_BUG_ON(new_port->port < (*list)->last->port);
+                    (*list)->last->next = new_port;
+                    new_port->prev = (*list)->last;
+                    (*list)->last = new_port;
+                } else {
+                    SCLogDebug("Port already exists in the list");
+                    goto error;
+                }
+            } else if (is_overlap && (new_port != NULL)) {
+                SCLogDebug("Found overlaps for [%u:%u], adding sigs", port, port2);
+                /* Only copy the relevant SGHs on later overlaps */
+                SigGroupHeadCopySigs(de_ctx, current->sh, &new_port->sh);
+            }
+            stack[stack_depth++] = current;
+            if (stack_depth == stack_size) {
+                SCLogDebug("Stack depth %d maxed out, realloc'ing..", stack_depth);
+                stack_size = 2 * stack_size;
+                void *tmp = SCRealloc(stack, stack_size * sizeof(SCPortIntervalNode *));
+                if (tmp == NULL) {
+                    SCLogError("Couldn't realloc the interval tree stack");
+                    goto error;
+                }
+                stack = tmp;
+            }
+            current = IRB_LEFT(current, irb);
+        }
+
+        if (stack_depth == 0) {
+            SCLogDebug("Stack depth was exhausted");
+            break;
+        }
+
+        SCPortIntervalNode *popped = stack[stack_depth - 1];
+        stack_depth--;
+        BUG_ON(popped == NULL);
+        current = IRB_RIGHT(popped, irb);
+    }
+    if (new_port != NULL)
+        SanitizePortList(de_ctx, list);
+    if (stack != NULL)
+        SCFree(stack);
+    return;
+error:
```

Comment:
This can happen in two scenarios:
1. memory failures
2. duplicate ports

In these cases, the caller skips adding the current port to the list and moves on to try adding the next one.